File: Smallest String Starting From Leaf.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        def exploreNode(node, path):
            if node is None:
                return path

            path.appendleft(node.val)
            if node.left and node.right:
                left_path = exploreNode(node.left, deque(path))
                right_path = exploreNode(node.right, deque(path))
                return min(left_path, right_path)
            elif node.left:
                return exploreNode(node.left, path)
            elif node.right:
                return exploreNode(node.right, path)
            else:
                return path

        arr = exploreNode(root, deque([]))
        n = len(arr)
        ans = []
        for i in range(n):
            if arr[i] >= 0:
                ans.append(chr(arr[i] + 97))
        return ''.join(ans)

File: Smallest String Starting From Leaf.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string smallestFromLeaf(TreeNode* root) {

        function<deque<int>(TreeNode*, deque<int>&)> exploreNode;
        exploreNode = [&exploreNode] (TreeNode* node, deque<int>& path) -> deque<int> {
            if (node == nullptr) return path;

            path.push_front(node->val);
            if (node->left && node->right) {
                deque<int> left_copy(path);
                deque<int> right_copy(path);
                auto left_path = exploreNode(node->left, left_copy);
                auto right_path = exploreNode(node->right, right_copy);
                return min(left_path, right_path);
            } else if (node->left) return exploreNode(node->left, path);
            else if (node->right) return exploreNode(node->right, path);
            else return path;
        };

        deque<int> q;
        auto arr = exploreNode(root, q);
        stringstream ss;
        for (int& num: arr) {
            ss << (char) (num+97);
        }
        return ss.str();
    }
};
